Fudan University Shanghai Cancer Center Triple-negative breast cancer (FUSCCTNBC) project is the largest single-center multi-omic cohort of triple-negative breast cancer to date (Jiang et al. Cancer Cell, 2019; 35[3]:428-440). Here we demonstrated how we used a modified version of the Characterization of Germline Variants (CharGer; Scott et al. Bioinformatics, 2018; 35[5]:865-867) to identify pathogenic/likely pathogenic variants in the FUSCCTNBC cohort. We have made two modifications to the standard pipeline of CharGer: First, as the current version of CharGer does not include minor allele frequency (MAF) data for the East Asian population, we manually filtered our candidate variants according to Genome Aggregation Database (gnomAD) (18). Only rare variants (MAF <0.5% in the East Asian population) were retained. Second, we manually updated the ClinVar records, and only variants known to be related to cancer were kept for PS1 and PM5 classifications.

We uploaded 3 de-identified vcfs of one TNBC patient (TNBC34_GATK.BamReadcount.MAFfiltered.VEP.vcf, TNBC34_Pindel.del_BamReadcount.MAFfiltered.VEP.vcf and TNBC34_Pindel.ins_BamReadcount.MAFfiltered.VEP.vcf) as examples. Preprocessings of the data include: 
1. Only variants with high confidence according to the following criteria were retained: 1) for SNVs, at least 20x coverage, allelic depth (AD) â‰¥ 5 for the alternative allele and 20% variant allele frequency (VAF) were required; 2) for indels, identified by both GATK and Pindel, or Pindel-unique calls with high confidence (at least 30x coverage and 20% VAF).
2. Only rare variants (MAF <0.5% in the East Asian population) were retained.


Note: Raw data of the FUSCCTNBC project can be viewed in The National Omics Data Encyclopedia (NODE) (http://www.biosino.org/node) by pasting the accession (OEP000155) into the text search box or through the URL: http://www.biosino.org/node/project/detail/OEP000155. Microarray data and sequence data have also been deposited in the NCBI Gene Expression Omnibus and Sequence Read Archive under accession numbers GSE118527 (OncoScan), and SRP157974 (WES and RNA-seq), respectively.
